README
NAME
    Acme::CXW::vars::i - Perl pragma to declare and simultaneously
    initialize global variables.

SYNOPSIS
        use Data::Dumper;
        $Data::Dumper::Deparse = 1;

        use vars::i '$VERSION' => 3.44;
        use vars::i '@BORG' => 6 .. 6;
        use vars::i '%BORD' => 1 .. 10;
        use vars::i '&VERSION' => sub(){rand 20};
        use vars::i '*SOUTH' => *STDOUT;

        BEGIN {
            print SOUTH Dumper [
                $VERSION, \@BORG, \%BORD, \&VERSION
            ];
        }

        use vars::i [ # has the same effect as the 5 use statements above
            '$VERSION' => 3.66,
            '@BORG' => [6 .. 6],
            '%BORD' => {1 .. 10},
            '&VERSION' => sub(){rand 20},
            '*SOUTH' => *STDOUT,
        ];

        print SOUTH Dumper [ $VERSION, \@BORG, \%BORD, \&VERSION ];

        __END__

DESCRIPTION
    For whatever reason, I once had to write something like

        BEGIN {
            use vars '$VERSION';
            $VERSION = 3;
        }

    and I really didn't like typing that much. With this package, I can say:

        use vars::i '$VERSION' => 3;

    and get the same effect.

    Also, I like being able to say

        use vars::i '$VERSION' => sprintf("%d.%02d", q$Revision: 1.3 $ =~ /: (\d+)\.(\d+)/);

        use vars::i [
         '$VERSION' => sprintf("%d.%02d", q$Revision: 1.3 $ =~ /: (\d+)\.(\d+)/),
         '$REVISION'=> '$Id: GENERIC.pm,v 1.3 2002/06/02 11:12:38 _ Exp $',
        ];

    Like with "use vars;", there is no need to fully qualify the variable
    name. However, you may if you wish.

NOTES
    *   Specifying a variable but not a value will succeed silently, and
        will not create the variable. E.g., "use vars::i '$foo';" is a
        no-op.

    *   Trying to create a special variable is fatal. E.g., "use vars::i
        '$@', 1;" will die at compile time.

SEE ALSO
    See vars, "our" in perldoc, "Pragmatic Modules" in perlmodlib.

AUTHORS
    D.H aka PodMaster, plus code from CXW.

    Please use http://rt.cpan.org/ to report bugs (there shouldn't be any
    ;p).

    Just go to http://rt.cpan.org/NoAuth/Bugs.html?Dist=vars-i to see a bug
    list and/or report new ones.

LICENSE
    Copyright (c) 2003 by D.H. aka PodMaster. Portions copyright (c) 2019 by
    Chris White. All rights reserved.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. If you don't know what this means,
    visit http://perl.com/ or http://cpan.org/.



Changes
Revision history for Perl distribution vars-i

{{$NEXT}}

2.000000 2019-09-18T01:21:48Z

    - Added a few more tests
    - Production release of 1.900000
    - Switch to Semantic Versioning

1.900000 2019-09-16T02:26:52Z (TRIAL RELEASE)

    - **Versioning change**: Prepare for switch to semantic
      versioning.  The next production release will be semver 2.0.0.
    - In `use vars::i '$foo'=>[1..3]`, make `$foo` hold
      an arrayref rather than the first element of the
      provided array.  (#4)
    - Fill in prerequisites and update documentation
    - Add test kit
    - Add a placeholder for options (should any be needed
      in the future).

1.10 2019-07-09T22:03:31Z
    - Production release

1.09 2019-07-08T14:47:55Z (TRIAL RELEASE)
    - Fixed typo in test files

1.08 2019-07-08T14:07:16Z (TRIAL RELEASE)
    - Support `use vars '%x' => <hashref>` (#2)
    - Support `use vars '@x' => <arrayref>`

1.07 2019-05-22T01:30:08Z
    - Updated documentation
    - Production release

1.06 2019-05-15T14:05:21Z (TRIAL RELEASE)
    - Fixed metadata

1.05 2019-05-14T13:39:22Z (TRIAL RELEASE)
    - Back under the vars::i name
    - Updated documentation

1.04 2019-04-26T20:49:06Z (TRIAL RELEASE)
    (as Acme::CXW::vars::i)
    - Switched to Minilla for maintenance

1.03 2019-03-27 (TRIAL RELEASE)
    (as Acme::CXW::vars::i)
    - Prohibit special names
    - 100% test coverage!

1.02 2019-03-25 (TRIAL RELEASE)
    (as Acme::CXW::vars::i)
    - Update to run on Perl 5.22+

1.01 2003-11-03
    - original version; created by ExtUtils::ModuleMaker 0.32
    - injected http://perlmonks.org/index.pl?node_id=264096
    - wrote tests
    - uploaded to cpan (finally)

# vi: set ts=4 sts=4 sw=4 et ai: #


t/Kit.pm
package #Hide from PAUSE
    Kit;
use 5.006001;
use strict;
use warnings;
use Import::Into;
use Test::More;

use Carp qw(croak);

use parent 'Exporter';

our @EXPORT;
BEGIN { @EXPORT=qw(eval_dies_ok eval_lives_ok eval_dies_like eval_is_var
    line_mark_string); }

=head1 NAME

Kit - vars::i test kit

=head1 SYNOPSIS

Test helpers.  These make it easier to trap undefined variables using string
eval and C<use strict>.

=head1 FUNCTIONS

=head2 eval_dies_ok

    eval_dies_ok "Code string" [, "message"];

Runs the code string; tests that the code died.
Any exception will be reported at the same line in the caller as the
C<eval_dies_ok> invocation.

=cut

sub eval_dies_ok {
    my (undef, $filename, $line) = caller;
    eval line_mark_string($filename, $line-1, $_[0]);
    ok($@, $_[1] || ('Died as expected: ' . $_[0]));
}

=head2 eval_lives_ok

    eval_lives_ok "Code string" [, "message"];

Runs the code string; tests that the code did not throw an exception.

=cut

sub eval_lives_ok {
    eval $_[0];
    is($@, '', $_[1] || ('Lived as expected: ' . $_[0]));
}

=head2 eval_dies_like

    eval_dies_like "Code string", qr/regex/ [, "message"];

Runs the code string; tests that the code threw an exception matching C<regex>.

=cut

sub eval_dies_like {
    eval $_[0];
    like($@, $_[1], $_[2] || ('Died with exception matching ' . $_[1]));
}

=head2 eval_is_var

    eval_is_var '$Package::var', value [, "message"];

Tests that C<$Package::var> exists, and that C<$Package::var eq value>.

=cut

sub eval_is_var {
    my ($varname, $expected, $msg) = @_;
    $msg ||= "$varname eq $expected";
    my ($sigil, $package, $basename) = ($varname =~ m/^(.)(.+)::([^:]+)$/);
    die "Invalid varname $varname" unless $package && $varname;

    my $got = eval qq[do { package $package; use strict; $sigil$basename }];

    is($@, '', "Accessed $varname");
    is($got, $expected, $msg);
} #eval_var_is

=head2 line_mark_string

Add a C<#line> directive to a string.  Usage:

    my $str = line_mark_string <<EOT ;
    $contents
    EOT

or

    my $str = line_mark_string __FILE__, __LINE__, <<EOT ;
    $contents
    EOT

In the first form, information from C<caller> will be used for the filename
and line number.

The C<#line> directive will point to the line after the C<line_mark_string>
invocation, i.e., the first line of <C$contents>.  Generally, C<$contents> will
be source code, although this is not required.

C<$contents> must be defined, but can be empty.

=cut

sub line_mark_string {
    my ($contents, $filename, $line);
    if(@_ == 1) {
        $contents = $_[0];
        (undef, $filename, $line) = caller;
    } elsif(@_ == 3) {
        ($filename, $line, $contents) = @_;
    } else {
        croak "Invalid invocation";
    }

    croak "Need text" unless defined $contents;
    die "Couldn't get location information" unless $filename && $line;

    $filename =~ s/"/-/g;
    ++$line;

    return <<EOT;
#line $line "$filename"
$contents
EOT
} #line_mark_string()

=head2 import

Exports the functions using L<Exporter> --- all functions are exported by
default.  Also loads L<Test::More> into the caller.

=cut

sub import {
    my $target = caller;
    __PACKAGE__->export_to_level(1, @_);
    Test::More->import::into($target);
    STDOUT->autoflush(1);
    STDERR->autoflush(1);
}


Makefile.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.006;
use strict;

use ExtUtils::MakeMaker 6.64;

my %WriteMakefileArgs = (
    NAME     => 'vars::i',
    DISTNAME => 'vars-i',
    VERSION  => '1.900000',
    EXE_FILES => [glob('script/*'), glob('bin/*')],
    CONFIGURE_REQUIRES => {
  "ExtUtils::MakeMaker" => "6.64"
}
,
    BUILD_REQUIRES     => {
  "ExtUtils::MakeMaker" => 0
}
,
    TEST_REQUIRES      => {
  "Exporter" => 0,
  "Import::Into" => "1.002005",
  "Test::More" => 0,
  "lib::relative" => "1.000",
  "parent" => 0
}
,
    PREREQ_PM          => {
  "Carp" => 0,
  "Data::Dumper" => 0,
  "perl" => "5.006001",
  "strict" => 0,
  "warnings" => 0
}
,
);

WriteMakefile(%WriteMakefileArgs);


lib/vars/i.pm
package vars::i;
use 5.006001;

our $VERSION = '1.900000';  # Prerelease leading to v2.0.0

use strict qw(vars subs);
use warnings;

# Turn a scalar, arrayref, or hashref into a list
sub _unpack {
    if( ref $_[0] eq 'ARRAY' ){
        return @{$_[0]};
    }
    elsif( ref $_[0] eq 'HASH' ){
        return %{$_[0]};
    }
    else {
        return ($_[0]);
    }
} #_unpack

sub import {
    return if @_ < 2;
    my( $pack, $first_var, @value ) = @_;
    my $callpack = caller;

    my %definitions;

    if( not @value ){
        if( ref $first_var eq 'ARRAY' ){  # E.g., use vars [ foo=>, bar=>... ];
            %definitions = @$first_var;
        }
        elsif( ref $first_var eq 'HASH' ){  # E.g., use vars { foo=>, bar=>... };
            %definitions = %$first_var;
        }
        else {
            return;     # No value given --- no-op; not an error.
        }
    }
    elsif(@value == 1) {        # E.g., use vars foo => <str/hashref/arrayref>
        %definitions = ( $first_var => $value[0] );
    }
    else {
        %definitions = ( $first_var => [@value] );
    }

    #require Data::Dumper;   # For debugging
    #print Data::Dumper->Dump([\%definitions], ['definitions']);

    while( my ($var, $val) = each %definitions ){

        if( my( $ch, $sym ) = $var =~ /^([-\$\@\%\*\&])(.+)$/ ){
            if( $ch eq '-' ){   # An option
                require Carp;
                Carp::croak('vars::i does not yet support any options!');
            }
            elsif( $sym !~ /^(\w+(::|'))+\w+$/ && $sym =~ /\W|(^\d+$)/ ){
                #    ^^ Skip fully-qualified names  ^^ Check special names

                # A variable name we can't or won't handle
                require Carp;

                if( $sym =~ /^\w+[[{].*[]}]$/ ){
                    Carp::croak("Can't declare individual elements of hash or array");
                }
                elsif( $sym =~ /^(\d+|\W|\^[\[\]A-Z\^_\?]|\{\^[a-zA-Z0-9]+\})$/ ){
                    Carp::croak("Refusing to initialize special variable $ch$sym");
                }
                else {
                    Carp::croak("I can't recognize $ch$sym as a variable name");
                }
            }

            $sym = "${callpack}::$sym" unless $sym =~ /::/;

            if( $ch eq '$' ){
                *{$sym} = \$$sym;
                ${$sym} = $val;
            }
            elsif( $ch eq '@' ){
                *{$sym} = \@$sym;
                @{$sym} = _unpack $val;
            }
            elsif( $ch eq '%' ){
                *{$sym} = \%$sym;
                %{$sym} = _unpack $val;
            }
            elsif( $ch eq '*' ){
                *{$sym} = \*$sym;
                (*{$sym}) = $val;
            }
            else {   # $ch eq '&'; guaranteed by the regex above.
                my ($param) = $val;
                if(ref $param) {
                    # NOTE: for now, permit any ref, since we can't determine
                    # refs overload &{}.  If necessary, we can later use
                    # Scalar::Util 1.25+'s blessed(), and allow CODE refs
                    # or blessed refs.
                    *{$sym} = $param;
                }
                else {
                    require Carp;
                    Carp::croak("Can't assign non-reference " .
                        (defined($param) ? $param : '<undef>') .
                        " to $sym");
                }
            }
            # There is no else, because the regex above guarantees
            # that $ch has one of the values we tested.

        }
        else {    # Name didn't match the regex above
            require Carp;
            Carp::croak("'$var' is not a valid variable or option name");
        }
    }
} #import()

1;
__END__

=head1 NAME

vars::i - Perl pragma to declare and simultaneously initialize global variables.

=head1 SYNOPSIS

    use Data::Dumper;
    $Data::Dumper::Deparse = 1;

    use vars::i '$VERSION' => 3.44;
    use vars::i '@BORG' => 6 .. 6;
    use vars::i '%BORD' => 1 .. 10;
    use vars::i '&VERSION' => sub(){rand 20};
    use vars::i '*SOUTH' => *STDOUT;

    BEGIN {
        print SOUTH Dumper [
            $VERSION, \@BORG, \%BORD, \&VERSION
        ];
    }

    use vars::i [ # has the same effect as the 5 use statements above
        '$VERSION' => 3.66,
        '@BORG' => [6 .. 6],
        '%BORD' => {1 .. 10},
        '&VERSION' => sub(){rand 20},
        '*SOUTH' => *STDOUT,
    ];

    print SOUTH Dumper [ $VERSION, \@BORG, \%BORD, \&VERSION ];

=head1 DESCRIPTION

For whatever reason, I once had to write something like

    BEGIN {
        use vars '$VERSION';
        $VERSION = 3;
    }

or

    our $VERSION;
    BEGIN { $VERSION = 3; }

and I really didn't like typing that much.  With this package, I can say:

    use vars::i '$VERSION' => 3;

and get the same effect.

Also, I like being able to say

    use vars::i '$VERSION' => sprintf("%d.%02d", q$Revision: 1.3 $ =~ /: (\d+)\.(\d+)/);

    use vars::i [
     '$VERSION' => sprintf("%d.%02d", q$Revision: 1.3 $ =~ /: (\d+)\.(\d+)/),
     '$REVISION'=> '$Id: GENERIC.pm,v 1.3 2002/06/02 11:12:38 _ Exp $',
    ];

Like with C<use vars;>, there is no need to fully qualify the variable name.
However, you may if you wish.

=head1 NOTES

=over

=item *

Specifying a variable but not a value will succeed silently, and will B<not>
create the variable.  E.g., C<use vars::i '$foo';> is a no-op.

Now, you might expect that C<< use vars::i '$foo'; >> would behave the same
way as C<< use vars '$foo'; >>.  That would not be an unreasonable expectation.
However, C<< use vars::i qw($foo $bar); >> has a very different
effect than does C<< use vars qw($foo $bar); >>!  In order to avoid
subtle errors in the two-parameter case, C<vars::i> also rejects the
one-parameter case.

=item *

Trying to create a special variable is fatal.  E.g., C<use vars::i '$@', 1;>
will die at compile time.

=item *

The sigil is taken into account (context sensitivity!)  So:

    use vars::i '$foo' => [1,2,3];  # now $foo is an arrayref
    use vars::i '@bar' => [1,2,3];  # now @bar is a three-element list

=back

=head1 SEE ALSO

See L<vars>, L<perldoc/"our">, L<perlmodlib/Pragmatic Modules>.

=head1 VERSIONING

Since version 1.900000, this module is numbered using
L<Semantic Versioning 2.0.0|https://semver.org>,
packed in the compatibility format of C<< vX.Y.Z -> X.00Y00Z >>.

This version supports Perl 5.6.1+.  If you are running an earlier Perl:

=over

=item Perl 5.6:

Use version 1.10 of this module
(L<CXW/vars-i-1.10|https://metacpan.org/pod/release/CXW/vars-i-1.10/lib/vars/i.pm>).

=item Pre-5.6:

Use version 1.01 of this module
(L<PODMASTER/vars-i-1.01|https://metacpan.org/pod/release/PODMASTER/vars-i-1.01/lib/vars/i.pm>).

=back

=head1 DEVELOPMENT

This module uses L<Minilla> for release management.  When developing, you
can use normal C<prove -l> for testing based on the files in C<lib/>.  Before
submitting a pull request, please:

=over

=item *

make sure all tests pass under C<minil test>

=item *

add brief descriptions to the C<Changes> file, under the C<{{$NEXT}}> line.

=item *

update the C<.mailmap> file to list your PAUSE user ID if you have one, and
if your git commits are not under your C<@cpan.org> email.  That way you will
be properly listed as a contributor in MetaCPAN.

=back

=head1 AUTHORS

D.H. <podmaster@cpan.org>

Christopher White <cxw@cpan.org>

=head2 Thanks

Thanks to everyone who has worked on L<vars>, which served as the basis for
this module.

=head1 SUPPORT

Please report any bugs at L<https://github.com/cxw42/Perl-vars-i/issues>.

You can also see the old bugtracker at
L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=vars-i> for older bugs.

=head1 LICENSE

Copyright (c) 2003--2019 by D.H. aka PodMaster, and contributors.
All rights reserved.

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
